servicetalk-encoding-api/src/main/java/io/servicetalk/encoding/api/AbstractZipContentCodec.java

```diff
@@ -204,87 +128,343 @@ public final Buffer decode(final Buffer src, final int offset, final int length,
             LOGGER.error("Error while decoding with {}", name(), e);
             throw new RuntimeException(e);
         } finally {
-            close(input);
+            closeIgnoreException(input);
         }
 
         return dst;
     }
 
     @Override
     public final Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator allocator) {
-        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+        return from.liftSync(subscriber -> new DecodingOperator(subscriber, allocator, this));
+    }
 
-            @Nullable
-            Inflater inflater;
-            @Nullable
-            ZLibStreamDecoder streamDecoder;
-            @Nullable
-            PublisherSource.Subscription subscription;
+    private static class EncodingOperator implements PublisherSource.Subscriber<Buffer> {
+
+        private static final int FOOTER_LEN = 10;
+        private static final OutputStream NULL_OUTPUT = newEmptyStream("NULL");
+        private static final OutputStream SIGNAL_IN_PROGRESS_OUTPUT = newEmptyStream("SIGNAL IN PROGRESS");
+        private static final OutputStream TERMINATED_OUTPUT = newEmptyStream("TERMINATED");
+
+        private static final AtomicLongFieldUpdater<EncodingOperator> demandUpdater =
+                newUpdater(EncodingOperator.class, "demand");
+
+        private static final AtomicReferenceFieldUpdater<EncodingOperator, OutputStream> outputUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(EncodingOperator.class, OutputStream.class, "output");
+
+        private final SwappableBufferOutputStream stream = new SwappableBufferOutputStream();
+        private final PublisherSource.Subscriber<? super Buffer> subscriber;
+        private final BufferAllocator allocator;
+        private final AbstractZipContentCodec codec;
+
+        private volatile long demand;
+
+        @Nullable
+        private volatile OutputStream output = NULL_OUTPUT;
+
+        EncodingOperator(final PublisherSource.Subscriber<? super Buffer> subscriber, final BufferAllocator allocator,
+                         final AbstractZipContentCodec codec) {
+            this.subscriber = subscriber;
+            this.allocator = allocator;
+            this.codec = codec;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            final Subscription sub = new Subscription() {
+                @Override
+                public void request(final long n) {
+                    if (!isRequestNValid(n)) {
+                        subscription.request(n);
+                    } else if (demandUpdater.accumulateAndGet(EncodingOperator.this, n,
+                            FlowControlUtils::addWithOverflowProtectionIfNotNegative) > 0) {
+                        subscription.request(n);
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    subscription.cancel();
```

Comment:
consider using `try/finally` to ensure cleanup code runs (same comment for DecodingOperator)